# Selection
Let's briefly go over how to use brackets for selection based off of comparison operators.

In [48]:
arr = np.arange(1,11)
arr

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [49]:
arr > 4

array([False, False, False, False,  True,  True,  True,  True,  True,
        True])

In [50]:
bool_arr = arr>4

In [51]:
bool_arr

array([False, False, False, False,  True,  True,  True,  True,  True,
        True])

In [52]:
arr[bool_arr]

array([ 5,  6,  7,  8,  9, 10])

In [53]:
arr[arr>2]

array([ 3,  4,  5,  6,  7,  8,  9, 10])

In [54]:
x = 2
arr[arr>x]

array([ 3,  4,  5,  6,  7,  8,  9, 10])

# NumPy Operations
Arithmetic
You can easily perform array with array arithmetic, or scalar with array arithmetic. Let's see some examples:

In [55]:
import numpy as np
arr = np.arange(0,10)

In [56]:
arr + arr

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [57]:
arr * arr

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

In [58]:
arr - arr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [59]:
# Warning on division by zero, but not an error!
# Just replaced with nan
arr/arr

C:\Users\admin\AppData\Local\Temp\ipykernel_14876\2878212635.py:3: RuntimeWarning: invalid value encountered in divide
  arr/arr


array([nan,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])

In [60]:
# Also warning, but not an error instead infinity
1/arr

C:\Users\admin\AppData\Local\Temp\ipykernel_14876\1360216608.py:2: RuntimeWarning: divide by zero encountered in divide
  1/arr


array([       inf, 1.        , 0.5       , 0.33333333, 0.25      ,
       0.2       , 0.16666667, 0.14285714, 0.125     , 0.11111111])

In [61]:
arr**3

array([  0,   1,   8,  27,  64, 125, 216, 343, 512, 729], dtype=int32)

# Universal Array Functions
Numpy comes with many universal array functions, which are essentially just mathematical operations you can use to perform the operation across the array. Let's show some common ones:

In [62]:
#Taking Square Roots
np.sqrt(arr)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ])

In [63]:
#Calcualting exponential (e^)
np.exp(arr)

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03])

In [64]:
np.max(arr) #same as arr.max()

9

In [65]:
np.sin(arr)


array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
       -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849])

In [66]:
np.log(arr)

C:\Users\admin\AppData\Local\Temp\ipykernel_14876\3120950136.py:1: RuntimeWarning: divide by zero encountered in log
  np.log(arr)


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458])

# NumPy Exercises
Now that we've learned about NumPy let's test your knowledge. We'll start off with a few simple tasks, and then you'll be asked some more complicated questions.

Import NumPy as np

In [68]:
import numpy as np 

 Create an array of 10 zeros

In [69]:
arr=np.ones(10)
arr

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

Create an array of 10 ones

In [71]:
arr=np.ones(10)
arr

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

Create an array of 10 fives

In [72]:
arr=np.ones(10)*5
arr

array([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.])

Create an array of the integers from 10 to 50

Create an array of the integers from 10 to 50

In [75]:
arr=np.arange(10,51)
arr

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49, 50])

Create an array of all the even integers from 10 to 50

In [76]:
arr=np.arange(10,51,2)
arr

array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
       44, 46, 48, 50])

Create a 3x3 matrix with values ranging from 0 to 8

In [78]:
arr=np.arange(0,9).reshape(3,3)
arr

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

Create a 3x3 identity matrix

In [80]:
arr=np.eye(3)
arr

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# random
Use NumPy to generate a random number between 0 and 1

In [81]:
import numpy as np 
from numpy.random import randn


In [82]:
arr =randn(1)
arr

array([-0.30329803])

Use NumPy to generate an array of 25 random numbers sampled from a standard normal distribution

In [83]:
a=np.random.randn(25)
a

array([-0.17947687,  0.99991378, -0.11230839,  0.82280102,  0.39167335,
       -0.04087085, -0.97742326, -0.41576347, -1.00792841, -1.50888936,
       -1.2262644 ,  0.21884957,  2.30462925,  0.62787038, -0.71281522,
       -0.37903332, -0.71412805,  0.4206816 , -1.00746884, -0.66148334,
       -1.09917404,  0.57653834, -0.16501008,  0.05461837,  0.92514791])

Create the following matrix:

array([[ 0.01,  0.02,  0.03,  0.04,  0.05,  0.06,  0.07,  0.08,  0.09,  0.1 ],
       [ 0.11,  0.12,  0.13,  0.14,  0.15,  0.16,  0.17,  0.18,  0.19,  0.2 ],
       [ 0.21,  0.22,  0.23,  0.24,  0.25,  0.26,  0.27,  0.28,  0.29,  0.3 ],
       [ 0.31,  0.32,  0.33,  0.34,  0.35,  0.36,  0.37,  0.38,  0.39,  0.4 ],
       [ 0.41,  0.42,  0.43,  0.44,  0.45,  0.46,  0.47,  0.48,  0.49,  0.5 ],
       [ 0.51,  0.52,  0.53,  0.54,  0.55,  0.56,  0.57,  0.58,  0.59,  0.6 ],
       [ 0.61,  0.62,  0.63,  0.64,  0.65,  0.66,  0.67,  0.68,  0.69,  0.7 ],
       [ 0.71,  0.72,  0.73,  0.74,  0.75,  0.76,  0.77,  0.78,  0.79,  0.8 ],
       [ 0.81,  0.82,  0.83,  0.84,  0.85,  0.86,  0.87,  0.88,  0.89,  0.9 ],
       [ 0.91,  0.92,  0.93,  0.94,  0.95,  0.96,  0.97,  0.98,  0.99,  1.  ]])

In [87]:
arr=np.arange(1,101).reshape(10,10)/100
arr

array([[0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 ],
       [0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 ],
       [0.21, 0.22, 0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 ],
       [0.31, 0.32, 0.33, 0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 ],
       [0.41, 0.42, 0.43, 0.44, 0.45, 0.46, 0.47, 0.48, 0.49, 0.5 ],
       [0.51, 0.52, 0.53, 0.54, 0.55, 0.56, 0.57, 0.58, 0.59, 0.6 ],
       [0.61, 0.62, 0.63, 0.64, 0.65, 0.66, 0.67, 0.68, 0.69, 0.7 ],
       [0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77, 0.78, 0.79, 0.8 ],
       [0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88, 0.89, 0.9 ],
       [0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99, 1.  ]])

Create an array of 20 linearly spaced points between 0 and 1:

In [88]:
arr=np.linspace(0,1,20)
arr

array([0.        , 0.05263158, 0.10526316, 0.15789474, 0.21052632,
       0.26315789, 0.31578947, 0.36842105, 0.42105263, 0.47368421,
       0.52631579, 0.57894737, 0.63157895, 0.68421053, 0.73684211,
       0.78947368, 0.84210526, 0.89473684, 0.94736842, 1.        ])

# Numpy Indexing and Selection
Now you will be given a few matrices, and be asked to replicate the resulting matrix outputs:

In [89]:
mat = np.arange(1,26).reshape(5,5)
mat

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [92]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[2:,1:]

array([[12, 13, 14, 15],
       [17, 18, 19, 20],
       [22, 23, 24, 25]])

In [96]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[3,4]

20

In [102]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[:3,1:2]

array([[ 2],
       [ 7],
       [12]])

In [105]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[-1]

array([21, 22, 23, 24, 25])

In [107]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[3:,]

array([[16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

# Now do the following

Get the sum of all the values in mat

In [109]:
mat.sum()

325

325

Get the standard deviation of the values in mat

In [111]:
mat.std()

7.211102550927978

7.2111025509279782

Get the sum of all the columns in mat

In [116]:
mat

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [117]:
mat.sum(axis=0)

array([55, 60, 65, 70, 75])

array([55, 60, 65, 70, 75])